infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/PrefixTrimRedundancyCalculator.java

```diff
@@ -0,0 +1,96 @@
+package org.corfudb.infrastructure.redundancy;
+
+import com.google.common.collect.ImmutableList;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Layout;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegment;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.NOT_TRANSFERRED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.RESTORED;
+import static org.corfudb.runtime.view.Address.NON_ADDRESS;
+
+/**
+ * A redundancy calculator that reads the global trim mark before creating a list of
+ * segments to transfer.
+ */
+public class PrefixTrimRedundancyCalculator extends RedundancyCalculator {
+
+    private final CorfuRuntime runtime;
+
+    public PrefixTrimRedundancyCalculator(String node, CorfuRuntime runtime) {
+        super(node);
+        this.runtime = runtime;
+    }
+
+    /**
+     * Sets the trim mark on this log unit and also performs a prefix trim.
+     *
+     * @param layout   A current layout.
+     * @param runtime  A current runtime.
+     * @param endpoint A current endpoint.
+     * @return A retrieved trim mark.
+     */
+    long setTrimOnNewLogUnit(Layout layout, CorfuRuntime runtime,
+                             String endpoint) {
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
+
+        Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
+        runtime.getLayoutView().getRuntimeLayout(layout)
+                .getLogUnitClient(endpoint)
+                .prefixTrim(prefixToken)
+                .join();
+        return trimMark;
+    }
+
+    /**
+     * Given a layout, creates an initial list of non-empty and bounded transfer segments.
+     *
+     * @param layout A current layout.
+     * @return A list of transfer segments.
+     */
+    public ImmutableList<TransferSegment> createStateList(Layout layout) {
+        long trimMark = setTrimOnNewLogUnit(layout, runtime, getServer());
```

Comment:
This is actually why I'm not convinced of the naming, I understand a calculator as some kind of static class which only computes something and returns. But we're even trimming the log from here. We could either review the abstraction or perhaps change the name?

Reply:
We can get rid of a PrefixTrim calculator, move createStateList in a regular calculator and call setTrimOnNewLogUnit as a part of a RestoreRedundancy action every time before we call create state list

Reply:
And trimMark now will be an argument to createStateList